src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -194,21 +194,56 @@ auto ClientRequest::reply_op_error(const Ref<PG>& pg, int err)
 }
 
 ClientRequest::interruptible_future<>
-ClientRequest::process_op(instance_handle_t &ihref, Ref<PG> &pg)
+ClientRequest::recover_missings(
+  instance_handle_t &ihref, Ref<PG> &pg)
 {
   return ihref.enter_stage<interruptor>(
-    client_pp(*pg).recover_missing,
-    *this
-  ).then_interruptible(
-    [this, pg]() mutable {
-    if (pg->is_primary()) {
-      return do_recover_missing(pg, m->get_hobj());
-    } else {
+    client_pp(*pg).recover_missing, *this
+  ).then_interruptible([this, pg]() mutable {
+    if (!pg->is_primary()) {
       logger().debug("process_op: Skipping do_recover_missing"
                      "on non primary pg");
       return interruptor::now();
     }
-  }).then_interruptible([this, pg, &ihref]() mutable {
+    auto fut = interruptor::now();
+    auto soid = m->get_hobj();
+    if (!soid.is_head()) {
+      fut = do_recover_missing(pg, soid.get_head());
+    }
+    return fut.then_interruptible([this, pg, soid] {
+      return pg->obc_loader.with_obc<RWState::RWREAD>(
+        soid.get_head(),
+        [this, pg, soid](auto head, auto) mutable {
+        auto oid = resolve_oid(head->get_head_ss(), soid);
+        assert(oid);
+        return do_recover_missing(pg, *oid
```

Comment:
I'd prefer to see this recover(head)->recover(snapshot) pattern moved into a helper.  Maybe rename do_recover_missing to recover_object and make do_recover_missing responsible for this part?

Reply:
Done.